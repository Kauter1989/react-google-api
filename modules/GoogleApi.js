import React from 'react'
import PropTypes from 'prop-types'
import { createContext } from 'react-broadcast'
import loadScript from '@lourd/load-script'

const {
  Provider: GoogleApiProvider,
  Consumer: GoogleApiConsumer,
} = createContext(null)
GoogleApiProvider.displayName = 'GoogleApiProvider'
GoogleApiConsumer.displayName = 'GoogleApiConsumer'

class GoogleApi extends React.Component {
  static propTypes = {
    clientId: PropTypes.string.isRequired,
    apiKey: PropTypes.string.isRequired,
    discoveryDocs: PropTypes.arrayOf(PropTypes.string).isRequired,
    scopes: PropTypes.arrayOf(PropTypes.string).isRequired,
    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  }

  authorize = () => {
    if (this.auth) {
      this.auth.signIn()
    }
  }

  signout = () => {
    if (this.auth) {
      this.auth.signOut()
    }
  }

  state = {
    signedIn: false,
    client: null,
    loading: true,
    error: null,
    authorize: this.authorize,
    signout: this.signout,
    googleAuth: this.auth,
  }

  componentDidMount() {
    this.setupApi()
  }

  async setupApi() {
    try {
      if (typeof window.gapi === 'undefined') {
        await loadScript('https://apis.google.com/js/api.js')
      }
      if (!gapi.client) {
        await new Promise((resolve, reject) =>
          gapi.load('client:auth2', {
            callback: resolve,
            onerror: reject,
          }),
        )
      }
      await gapi.client.init({
        apiKey: this.props.apiKey,
        clientId: this.props.clientId,
        discoveryDocs: this.props.discoveryDocs,
        scope: this.props.scopes.join(','),
      })
    } catch (error) {
      this.setState({
        loading: false,
        error,
      })
      return
    }
    this.auth = gapi.auth2.getAuthInstance()
    this.setState({
      client: gapi.client,
      loading: false,
      signedIn: this.auth.isSignedIn.get(),
      googleAuth: this.auth,
    })
    this.auth.isSignedIn.listen(signedIn => this.setState({ signedIn }))
  }

  render() {
    return (
      <GoogleApiProvider value={this.state}>
        {typeof this.props.children === 'function'
          ? this.props.children(this.state)
          : this.props.children}
      </GoogleApiProvider>
    )
  }
}

export { GoogleApi, GoogleApiConsumer, GoogleApiProvider }