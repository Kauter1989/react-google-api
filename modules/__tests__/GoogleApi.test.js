import React from 'react'
import { mount as render } from 'enzyme'
import loadScript from '@lourd/load-script'
import Catcher from '@lourd/react-catcher'

// importing from the public API
import { GoogleApi, GoogleApiConsumer } from '../'
import { GoogleApiProvider } from '../GoogleApi'

jest.mock('@lourd/load-script')

const apiProps = {
  clientId: 'id',
  apiKey: 'apiKey',
  discoveryDocs: ['strings'],
  scopes: ['more strings', 'many strings'],
}

describe('GoogleApi component', () => {
  it(`has initial values for the api state, including the bound authorize and
  signout instance methods, and passes that state down through a context
  provider and the children prop if it is a function`, () => {
    // mock out side effect function
    const spy = jest.spyOn(GoogleApi.prototype, 'setupApi').mockImplementation()
    const mockChildrenFn = jest.fn(api => {
      expect(api).toMatchSnapshot()
      return 'foo'
    })
    const result = render(<GoogleApi {...apiProps}>{mockChildrenFn}</GoogleApi>)
    expect(result.is(GoogleApiProvider))
    const contextValue = result.children().prop('value')
    const state = result.state()
    expect(contextValue).toBe(state)
    const instance = result.instance()
    expect(instance.authorize).toBe(state.authorize)
    expect(instance.signout).toBe(state.signout)
    const childFunctionArg = mockChildrenFn.mock.calls[0][0]
    expect(contextValue).toBe(childFunctionArg)
    expect(GoogleApi.prototype.setupApi).toHaveBeenCalled()
    spy.mockRestore()
  })

  it(`handles authorize or signout being called when the Google API client has
  has not been setup yet`, () => {
    const spy = jest.spyOn(GoogleApi.prototype, 'setupApi').mockImplementation()
    const instance = render(
      <GoogleApi {...apiProps}>{null}</GoogleApi>,
    ).instance()
    expect(instance.auth).toBe(undefined)
    instance.authorize()
    instance.signout()
    spy.mockRestore()
  })

  it(`Throws a render error when no children are given`, async () => {
    const setup = () => {
      const spy = jest
        .spyOn(GoogleApi.prototype, 'setupApi')
        .mockImplementation()
      // suppress react's console warnings
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation()
      return () => {
        spy.mockRestore()
        consoleSpy.mockRestore()
      }
    }
    const cleanup = setup()
    const fn = jest.fn()
    render(
      <Catcher onCatch={fn}>
        <GoogleApi {...apiProps} />
      </Catcher>,
    )
    expect(fn).toMatchSnapshot('no children render error')
    // being certain that it's only the two react errors being logged
    expect(console.error).toHaveBeenCalledTimes(2)
    cleanup()
  })

  describe('setupApi', () => {
    it(`loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state`, async () => {
      const mockAuth = {
        isSignedIn: { listen: jest.fn(), get: jest.fn(() => false) },
      }
      const mockSecondApiLoadStep = api => {
        api.client = { init: jest.fn() }
        api.auth2 = { getAuthInstance: jest.fn(() => mockAuth) }
      }
      const mockApi = () => {
        const gapi = {}
        gapi.load = jest.fn((string, opts) => {
          mockSecondApiLoadStep(gapi)
          opts.callback()
        })
        global.gapi = gapi
        return () => {
          delete global.gapi
        }
      }
      const setupMockEffects = () => {
        let cleanupApi
        loadScript.mockImplementation(() => {
          cleanupApi = mockApi()
        })
        return () => {
          loadScript.mockRestore()
          cleanupApi()
        }
      }
      const authSetter = jest.fn()
      const context = { setState: jest.fn(), props: apiProps }

      const cleanup = setupMockEffects()
      await GoogleApi.prototype.setupApi.call(context)
      expect(loadScript).toMatchSnapshot('loadScript call')
      expect(global.gapi.load).toMatchSnapshot('gapi.load call')
      expect(global.gapi.client.init).toMatchSnapshot('client init call')
      expect(context.setState).toMatchSnapshot('setState call')
      expect(context.auth).toBe(mockAuth)
      const authChangeSubscriber = mockAuth.isSignedIn.listen.mock.calls[0][0]
      authChangeSubscriber(true)
      expect(context.setState).toMatchSnapshot(
        'setState call from state change',
      )
      cleanup()
    })

    it(`catches errors from loading the api library and sets it in state`, async () => {
      const err = new Error('faiiiiiil')
      const setup = () => {
        loadScript.mockImplementation(() => Promise.reject(err))
        return () => {
          loadScript.mockRestore()
        }
      }
      const cleanup = setup()
      const context = { setState: jest.fn(), props: apiProps }
      await GoogleApi.prototype.setupApi.call(context)
      expect(context.setState).toMatchSnapshot(
        'library load error setState call',
      )
      cleanup()
    })

    it(`catches errors from loading the api auth extension and sets it in state`, async () => {
      const err = new Error('faiiiiiil')
      const mockApi = () => {
        const gapi = {}
        gapi.load = jest.fn((string, opts) => {
          opts.onerror(err)
        })
        global.gapi = gapi
        return () => {
          delete global.gapi
        }
      }
      const setupMockEffects = () => {
        let cleanupApi
        loadScript.mockImplementation(() => {
          cleanupApi = mockApi()
        })
        return () => {
          loadScript.mockRestore()
          cleanupApi()
        }
      }
      const cleanup = setupMockEffects()
      const context = { setState: jest.fn(), props: apiProps }
      await GoogleApi.prototype.setupApi.call(context)
      expect(context.setState).toMatchSnapshot('auth load error setState call')
      cleanup()
    })

    it(`catches errors from initializing the api client and sets it in state`, async () => {
      const err = new Error('faiiiiiil')
      const mockApi = () => {
        const mockClient = obj => {
          obj.client = {
            init: jest.fn(() => Promise.reject(err)),
          }
        }
        const gapi = {}
        gapi.load = jest.fn((string, opts) => {
          mockClient(gapi)
          opts.callback()
        })
        global.gapi = gapi
        return () => {
          delete global.gapi
        }
      }
      const setupMockEffects = () => {
        let cleanupApi
        loadScript.mockImplementation(() => {
          cleanupApi = mockApi()
        })
        return () => {
          loadScript.mockRestore()
          cleanupApi()
        }
      }
      const cleanup = setupMockEffects()
      const context = { setState: jest.fn(), props: apiProps }
      await GoogleApi.prototype.setupApi.call(context)
      expect(context.setState).toMatchSnapshot(
        'client init error setState call',
      )
      cleanup()
    })
  })
})

describe('GoogleApiConsumer component', () => {
  it(`gets the GoogleApi state as a render prop passed to the children
  function prop`, () => {
    // mock out side effect function
    const spy = jest.spyOn(GoogleApi.prototype, 'setupApi').mockImplementation()
    const mockContextFn = jest.fn(api => {
      expect(api).toMatchSnapshot()
      return 'foo'
    })
    const result = render(
      <GoogleApi {...apiProps}>
        <div>
          <GoogleApiConsumer>{mockContextFn}</GoogleApiConsumer>
        </div>
      </GoogleApi>,
    )
    const apiState = result.state()
    const contextRenderProp = mockContextFn.mock.calls[0][0]
    expect(contextRenderProp).toBe(apiState)
    spy.mockRestore()
  })
})
