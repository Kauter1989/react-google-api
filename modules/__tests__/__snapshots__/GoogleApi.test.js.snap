// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleApi component Throws a render error when no children are given: no children render error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      [Invariant Violation: GoogleApiProvider(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.],
      Object {
        "componentStack": "
    in GoogleApiProvider (created by GoogleApi)
    in GoogleApi
    in Catcher (created by WrapperComponent)
    in WrapperComponent",
      },
    ],
  ],
}
`;

exports[`GoogleApi component has initial values for the api state, including the bound authorize and
  signout instance methods, and passes that state down through a context
  provider and the children prop if it is a function 1`] = `
Object {
  "authorize": [Function],
  "client": null,
  "error": null,
  "loading": true,
  "signedIn": false,
  "signout": [Function],
}
`;

exports[`GoogleApi component setupApi catches errors from initializing the api client and sets it in state: client init error setState call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "error": [Error: faiiiiiil],
        "loading": false,
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi catches errors from loading the api auth extension and sets it in state: auth load error setState call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "error": [Error: faiiiiiil],
        "loading": false,
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi catches errors from loading the api library and sets it in state: library load error setState call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "error": [Error: faiiiiiil],
        "loading": false,
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state: client init call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "apiKey": "apiKey",
        "clientId": "id",
        "discoveryDocs": Array [
          "strings",
        ],
        "scope": "more strings,many strings",
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state: gapi.load call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "client:auth2",
      Object {
        "callback": [Function],
        "onerror": [Function],
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state: loadScript call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://apis.google.com/js/api.js",
    ],
  ],
}
`;

exports[`GoogleApi component setupApi loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state: setState call 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "client": Object {
          "init": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "apiKey": "apiKey",
                  "clientId": "id",
                  "discoveryDocs": Array [
                    "strings",
                  ],
                  "scope": "more strings,many strings",
                },
              ],
            ],
          },
        },
        "loading": false,
        "signedIn": false,
      },
    ],
  ],
}
`;

exports[`GoogleApi component setupApi loads the google api library, loads the client:auth2 extension, gets
    the auth instance, updates the component state, and subscribes to changes
    in the api client's internal auth state: setState call from state change 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "client": Object {
          "init": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "apiKey": "apiKey",
                  "clientId": "id",
                  "discoveryDocs": Array [
                    "strings",
                  ],
                  "scope": "more strings,many strings",
                },
              ],
            ],
          },
        },
        "loading": false,
        "signedIn": false,
      },
    ],
    Array [
      Object {
        "signedIn": true,
      },
    ],
  ],
}
`;

exports[`GoogleApiConsumer component gets the GoogleApi state as a render prop passed to the children
  function prop 1`] = `
Object {
  "authorize": [Function],
  "client": null,
  "error": null,
  "loading": true,
  "signedIn": false,
  "signout": [Function],
}
`;
